### PR TITLE
Jhbate/lastupload

### PR DESCRIPTION
**Pull request by jh-bate**
@darinkrauss if you could take an initial look at this it would be appreciated. I am still doing testing locally but keen to get your initial input 

Goals were:
 - make the ``/upload/lastentry/{userID}`` and  ``/upload/lastentry/{userID}/{deviceID}`` useable so we can better query accounts to find their state - i.e. do they have data and when was it last uploaded

 - a secondary goal was to cleanup the api code into one file for easier maintenance and understand for when we migrate to the standardised download api

### CODE REVIEW

File: clients/mongoStoreClient_test.go [hunk 1]
```diff
@@ -171,10 +189,57 @@ func TestIndexes(t *testing.T) {
...
+	}
+
+	if err != nil {
+		t.Fatalf("GetTimeLastEntryUserAndDevice unexpected error [%s]", err.Error())
+	}
 
 }
 
-func TestWhereQueryConstruction(t *testing.T) {
+func Test_constructQuery_WhereQueryConstruction(t *testing.T) {
```
Comment (darinkrauss) [inline clients/mongoStoreClient_test.go]: Was this rename intentional (or the result of a find/replace)?

File: clients/mongoStoreClient.go [hunk 2]
```diff
@@ -82,36 +83,51 @@ func (d MongoStoreClient) Ping() error {
...
-	fullQuery := bson.M{"$and": []bson.M{groupIdQuery, deviceIdQuery}}
-	// Get the entry with the latest time by reverse sorting and taking the first value
-	d.deviceDataC.Find(fullQuery).Sort("-time").One(&result)
-	bytes, err := json.Marshal(result["time"])
+	startQueryTime := time.Now()
+	sessionCopy := d.session
+	defer sessionCopy.Close()
+
+	err := sessionCopy.DB("").C(DEVICE_DATA_COLLECTION).
+		Find(bson.M{"$and": []bson.M{getBaseQuery(groupId), bson.M{"deviceId": deviceId}}}).
```
Comment (darinkrauss) [inline clients/mongoStoreClient.go]: Did you purposefully drop off the "or groupId" from the query?

Comment (jh-bate) [inline clients/mongoStoreClient.go]: yes - we aren't concerned with the old data model records for this I feel

File: clients/mongoStoreClient_test.go (hunk 1 above)
Comment (jh-bate) [inline clients/mongoStoreClient_test.go]: intentional - bad probably still bad. Its testing the internal method constructQuery with a where. I will have another look and rename as it still doesn't read as I want

File: api/octopusApi.go [hunk 3]
```diff
@@ -150,42 +151,163 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
...
+		query := string(rawQuery)
+
+		log.Println(QUERY_API_PREFIX, "Query: raw ", query)
+
+		errs, qd := model.BuildQuery(query)
+
+		if len(errs) != 0 {
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("Query: errors [%v] found parsing raw query [%s]", errs, query))
+			log.Println(QUERY_API_PREFIX, fmt.Sprintf("Query: failed after [%.5f] secs", time.Now().Sub(start).Seconds()))
+			http.Error(res, fmt.Sprintf("Errors building query: [%v]", errs), http.StatusBadRequest)
```
Comment (darinkrauss) [inline api/octopusApi.go]: Might want to put the errs array out as JSON rather than Go format for parsability?

**Comment by darinkrauss** [comment]
From what I can tell it looks good.

File: api/octopusApi.go (hunk 3 above)
Comment (jh-bate) [inline api/octopusApi.go]: true - will update but maybe overly ambitious at this stage

**Comment by jh-bate** [comment]
@darinkrauss made some changes to include the jsonError stuff here also.
 
I will create an issue in go-common about having common error stuff residing there and then we can also start working on standard errors e.g. not authorised, no permission to upload etc

**Comment by darinkrauss** [comment]
Looks good. :shipit:
